src/expr/term_conversion_proof_generator.cpp

```diff
@@ -18,27 +18,57 @@ using namespace CVC4::kind;
 
 namespace CVC4 {
 
+std::ostream& operator<<(std::ostream& out, TConvPolicy tcpol)
+{
+  switch (tcpol)
+  {
+    case TConvPolicy::FIXPOINT: out << "FIXPOINT"; break;
+    case TConvPolicy::ONCE: out << "ONCE"; break;
+    default: out << "TConvPolicy:unknown"; break;
+  }
+  return out;
+}
+
 TConvProofGenerator::TConvProofGenerator(ProofNodeManager* pnm,
-                                         context::Context* c)
-    : d_proof(pnm, nullptr, c), d_rewriteMap(c ? c : &d_context)
+                                         context::Context* c,
+                                         TConvPolicy tpol)
+    : d_proof(pnm, nullptr, c), d_rewriteMap(c ? c : &d_context), d_policy(tpol)
 {
 }
 
 TConvProofGenerator::~TConvProofGenerator() {}
 
 void TConvProofGenerator::addRewriteStep(Node t, Node s, ProofGenerator* pg)
 {
-  // should not rewrite term more than once
-  Assert(!hasRewriteStep(t));
+  if (t == s)
+  {
+    // nothing to do
+    return;
+  }
+  // should not rewrite term to two different things
+  if (hasRewriteStep(t))
+  {
+    Assert(getRewriteStep(t) == s);
+    return;
+  }
   Node eq = t.eqNode(s);
   d_proof.addLazyStep(eq, pg);
   d_rewriteMap[t] = s;
```

Comment:
You could encapsulate all but the step addition to `d_proof` in a `addRewriteStep` method that returns a boolean and is called by all three versions of `addRewriteStep`.